include/swift/AST/DiagnosticConsumer.h

```diff
@@ -209,16 +286,15 @@ class FileSpecificDiagnosticConsumer : public DiagnosticConsumer {
   /// Xcode will only see an error for a particular primary in that primary's
   /// serialized diagnostics file. So, tell the subconsumers to inform the
   /// driver of incomplete batch mode compilation.
-  void tellSubconsumersToInformDriverOfIncompleteBatchModeCompilation() const;
+  void tellSubconsumersToInformDriverOfIncompleteBatchModeCompilation();
 
   void computeConsumersOrderedByRange(SourceManager &SM);
 
   /// Returns nullptr if diagnostic is to be suppressed,
   /// None if diagnostic is to be distributed to every consumer,
   /// a particular consumer if diagnostic goes there.
-  Optional<ConsumerSpecificInformation *>
-  consumerSpecificInformationForLocation(SourceManager &SM,
-                                         SourceLoc loc) const;
+  Optional<ConsumerAndRange *> consumerAndRangeForLocation(SourceManager &SM,
```

Comment:
I think at this point you could go all the way to `Optional<Subconsumer *>`, since there's no more interesting information in ConsumerAndRange.

Reply:
It's a good idea, but runs into `const` trouble. The `[]` operator on `FileSpecificDiagnosticConsumer` is not `const` because we may be setting the emitted error bit in the subconsumer structure. `consumerAndRangeForLocation` is `const`. So I can see these choices:

1. `const_cast` `this` at the end of `consumerAndRangeForLocation`,
2. unconstify `consumerAndRangeForLocation`
3. let the `[]` operator be `const` and `const_cast` it to set the error bit.
4. status quo

What do you like? Or is there another choice?

Reply:
Ah, hm. I guess (2) is my preferred choice, then, because while `consumerAndRangeForLocation` isn't doing any mutation itself it's making it possible for callers to mutate. I see how (4) avoids that problem, but it does so at the cost of making the ConsumerAndRange structure more prominent and making things less convenient for callers. But if you have another preference that's okay; I see the difficulty now.

Reply:
Implemented (2)--looks good!